File: 0322-coin-change/0322-coin-change.java
class Solution {
    private int solve(int[] coins, int target, Map<Integer, Integer> dp) {
        if (target < 0)
            return Integer.MAX_VALUE;
        if (target == 0)
            return 0;

        if (dp.containsKey(target))
            return dp.get(target);
        int minCoins=Integer.MAX_VALUE;
        for (int coin : coins) {
            int newTarget = target - coin;

            int ans = solve(coins, newTarget, dp) ;

            if(ans!=Integer.MAX_VALUE)
            {
                minCoins=Math.min(ans+1,minCoins);
            }

        }
        dp.put(target, minCoins);
        return minCoins;
    }

    public int coinChange(int[] coins, int amount) {
        Map<Integer, Integer> dp = new HashMap<>();
        int ans = solve(coins, amount, dp);
        if(ans==Integer.MAX_VALUE)
        return -1;
        return ans;
    }
}

File: 0322-coin-change/0322-coin-change.py
class Solution:

    def solve(self,coins,amount,memo):
        # print("checking for ",amount,memo)
        if amount ==0:
            return 0
        if amount in memo :
            return memo[amount]
        memo[amount]=sys.maxsize
        for coin in coins:
            target=amount-coin
            if target>=0:
                ans=self.solve(coins,target,memo)+1
                memo[amount]=min(ans,memo[amount])
        return memo[amount]

    def coinChange(self, coins: List[int], amount: int) -> int:
        ans= self.solve(coins,amount,{})
        if ans==sys.maxsize:
            return -1 
        
        return ans
        